docs/architecture/architecture.md

```diff
@@ -0,0 +1,671 @@
+
+![Docker start](arch-images/start.png)
+
+#### [`exec`](https://github.com/clearcontainers/runtime/blob/master/exec.go)
+
+OCI `exec` allows you to run an additional command within an already running
+container. With Clear Containers, this translates into sending a `EXECCMD` command
+to the agent so that it runs a command into a running container belonging to a
+certain pod. All I/O streams from the executed command will be passed back to
+`containerd` or `conmon` through a newly created `cc-shim`.
+
+The `exec` code path is partly similar to the `create` one and `cc-runtime`
+goes through the following steps:
+
+1. `cc-runtime` connects to `cc-proxy` and sends it the `attach` command to let
+  it know which VM to run the `exec` command.
+2. `cc-runtime` receives a token from `cc-proxy` based on this connection.
+3. `cc-runtime` spawns a `cc-shim` process providing two arguments:
+  `cc-shim --token $(token) --uri $(uri)`
+   * The proxy URL, which can be either a UNIX or a TCP socket.
+   * The token for the container process it needs to monitor.
+  The `cc-shim` process will forward the output streams (stderr and stdout) to
+  either `containerd-shim` in the Docker use cases or `conmon` for the Kubernetes
+  deployements.
+4. `cc-runtime` sends an agent `EXECMD` command to start the command in the
+  right container. The command is sent to `cc-proxy` who forwards it to the right
+  agent instance running in the appropriate guest.
+
+Now the `exec`'ed process is running in the virtual machine, sharing the UTS,
+PID, mount and IPC namespaces with the container's init process.
+
+#### [`kill`](https://github.com/clearcontainers/runtime/blob/master/kill.go)
+
+When sending the OCI `kill` command, container runtimes should send a [UNIX signal](https://en.wikipedia.org/wiki/Unix_signal)
+to the container process.
+In the Clear Containers context, this means `cc-runtime` needs to handle 2 resources:
+
+1. It needs to send the signal to the container running on the VM.
+2. It needs to signal the [`shim`](#shim) since we don't want to leave dangling
+  `cc-shim` instances on the host in case the container process terminates inside
+  the VM.
+
+To do so, `cc-runtime` goes through the following steps:
+
+1. `cc-runtime` connects to `cc-proxy` and sends it the `attach` command to let
+  it know on which VM the container it is trying to `kill` is running.
+2. `cc-runtime` sends an agent `KILLCONTAINER` command to `kill` the container
+  running on the guest. The command is sent to `cc-proxy` who forwards it to the
+  right agent instance running in the appropriate guest.
+
+After step #2, if the container process terminates in the VM, `cc-proxy` will
+forward the process exit code to the `shim`. The `shim` will then disconnect
+from `cc-proxy` and terminates as well.
+
+One corner case `cc-runtime` also needs to handle is when it receives a signal
+for a container that's not yet running. In that case, `cc-runtime` simply stops
```

Comment:
This reads a little strange to me. Let's see if I can clarify (and somebody might want to check it is still accurate...)
'In that case, for the `SIGTERM` or `SIGKILL` signals, `cc-runtime` will kill the `cc-shim`. All other signals are ignored.'

Reply:
this reads much better to me, good suggestion.